### ISSUE TITLE
Configuring tests with ignoreFailures=true results in test failures being cached in the Build Cache

### ISSUE THREAD

**Issue by akomakom** [gradle/gradle#13893]
Ordinarily the Build Cache is not populated with test task outputs when the gradle build fails.  However, in CI, we ignore failures in order to disambiguate `unstable` (tests failed) and `failed` (any other reason for exit code > 0).  In Jenkins, this is a yellow vs a red build.  I understand that it is a common practice to have CI determine whether tests failed by inspecting junit xml:

```groovy
// Excerpt from an init plugin we use in CI:
project.plugins.withId('java') {
    project.test {
        ignoreFailures = true
    }
}
```
Now, running gradle with build cache turned on (we use a remote build cache) results in the failed test being cached:
```./gradlew test --build-cache```  

The test may have failed due to environment-related issues, but it will not rerun without `--rerun-tasks`, something that isn't practical in CI (without manual intervention).  My expectation would be that the failed test should always re-run.

In my opinion, no task output should be cached on failure.  
I've attempted to code around this specific instance as follows (unsuccessfully):

```groovy
project.plugins.withId('java') {
    project.test {
        ignoreFailures = true

        def failedTestCount = 0

        afterSuite { TestDescriptor desc, TestResult result ->
            failedTestCount += result.failedTestCount
        }

        outputs.doNotCacheIf("Do not cache test failures", new Spec<Task>() {
            @Override
            public boolean isSatisfiedBy(Task element) {
                return failedTestCount > 0
            }
        });
    }
}

```
It turns out that it is not possible to configure caching after the task started executing, and `afterSuite` runs after `isSatisfiedBy`.

In other words, my only option is to disable test output caching altogether.

**Is there a better way?**

**Gradle**: 6.5.1
